x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/frequentitemsets/EclatMapReducer.java

```diff
@@ -0,0 +1,368 @@
+import java.util.stream.Stream;
+
+/**
+ * Map Reducer that implements the eclat algorithm, see
+ *
+ * Zaki et.al. 1997 New Algorithms for Fast Discovery of Association Rules KDD'97, 283-296
+ * Christian Borgelt 2003 Efficient Implementations of Apriori and Eclat FIMI 2003
+ *
+ * Eclat is a successor of the well known apriori algorithm, in contrast to apriori eclat
+ * mines for item sets depth first instead of breath first. The main advantage is the lower
+ * memory requirements. Depth-first only requires state with a small footprint in contrast to
+ * a depth first approach which has to remember tons of sets in parallel.
+ *
+ * The implementation contains several adjustments and enhancements:
+ *
+ *  - top-N retrieval via the `size` parameter, this has a significant influence on runtime
+ *  - multi field support, the classic apriori has only a bag of items of the same kind
+ *  - mapping and merging results per shard
+ *
+ * Algorithmic differences:
+ *
+ *  - pruning/rewriting of items/transaction prior mining
+ *  - a fixed size bitvector to speed up traversal for known branches (C. Borgelt uses a similar technique)
+ *  - adaptive pruning based on a changing minimum support (for top-N retrieval)
+ */
+public class EclatMapReducer implements MapReducer {
+    private static final Logger logger = LogManager.getLogger(EclatMapReducer.class);
+
+    private static final int VERSION = 1;
+    public static final String NAME = "frequent_items-eclat-" + VERSION;
+
+    // cache for marking transactions visited, memory usage: ((BITSET_CACHE_TRAVERSAL_DEPTH -2) * BITSET_CACHE_NUMBER_OF_TRANSACTIONS) / 8
+    private static final int MAX_BITSET_CACHE_NUMBER_OF_TRANSACTIONS = 65536;
+    private static final int BITSET_CACHE_TRAVERSAL_DEPTH = 20;
+
+    private final String aggregationWritableName;
+
+    private final double minimumSupport;
+    private final int minimumSetSize;
+    private final int size;
+
+    private Iterable<FrequentItemSetCollector.FrequentItemSet> frequentSets = null;
+    private TransactionStore transactionStore;
+    private long eclatRuntimeNanos = 0;
+
+    public EclatMapReducer(String aggregationWritableName, double minimumSupport, int minimumSetSize, int size) {
+        this.aggregationWritableName = aggregationWritableName;
+        this.minimumSupport = minimumSupport;
+        this.minimumSetSize = minimumSetSize;
+        this.size = size;
+
+        // TODO: big arrays should be taken from AggregationContext
+        this.transactionStore = new TransactionStore(BigArrays.NON_RECYCLING_INSTANCE);
+    }
+
+    public EclatMapReducer(StreamInput in) throws IOException {
+        this.aggregationWritableName = in.readString();
+
+        // parameters
+        this.minimumSupport = in.readDouble();
+        this.minimumSetSize = in.readInt();
+        this.size = in.readInt();
+
+        // data
+        this.transactionStore = new TransactionStore(in, BigArrays.NON_RECYCLING_INSTANCE);
```

Comment:
the constructor `Something(StreamInput in)` is dictated by the the framework, I will try to find a way to workaround, but this might turn out to be impossible without a larger refactoring.